packages/firestore/src/local/indexeddb_persistence.ts

```diff
@@ -132,23 +170,143 @@ export class IndexedDbPersistence implements Persistence {
     assert(!this.started, 'IndexedDbPersistence double-started!');
     this.started = true;
 
+    assert(
+      this.window !== null && this.document !== null,
+      "Expected 'window' and 'document' to be defined"
+    );
+
     return SimpleDb.openOrCreate(this.dbName, SCHEMA_VERSION, createOrUpgradeDb)
       .then(db => {
         this.simpleDb = db;
       })
-      .then(() => this.tryAcquireOwnerLease())
       .then(() => {
-        this.scheduleOwnerLeaseRefreshes();
+        this.attachVisibilityHandler();
         this.attachWindowUnloadHook();
+        return this.schedulePrimaryLeaseRefreshes(0);
       });
   }
 
+  setPrimaryStateListener(primaryStateListener: PrimaryStateListener) {
+    this.primaryStateListener = primaryStateListener;
+    primaryStateListener(this.isPrimary);
+  }
+
+  /**
+   * Schedules a recurring timer to refresh or acquire the primary lease.
+   */
+  private schedulePrimaryLeaseRefreshes(initialDelayMs: number): Promise<void> {
+    return this.queue.enqueueAfterDelay(
+      TimerId.ClientStateRefresh,
+      initialDelayMs,
+      () => {
+        return this.tryBecomePrimary().then(currentPrimary => {
+          let nextDelayMs;
+
+          if (this.isLocalClient(currentPrimary)) {
+            nextDelayMs = PRIMARY_LEASE_MAX_REFRESH_INTERVAL_MS;
+          } else if (currentPrimary !== null) {
+            nextDelayMs = Math.max(
+              PRIMARY_LEASE_MIN_REFRESH_INTERVAL_MS,
+              Date.now() -
+                (currentPrimary.leaseTimestampMs + CLIENT_STATE_MAX_AGE_MS) +
+                1
+            );
+          } else {
+            nextDelayMs = PRIMARY_LEASE_MIN_REFRESH_INTERVAL_MS;
+          }
+
+          this.schedulePrimaryLeaseRefreshes(nextDelayMs);
+        });
+      }
+    );
+  }
+
+  /**
+   * Attempts to obtain the primary lease for the local client. If successful,
+   * returns a `DbOwner` with the local client id. Otherwise, returns the
+   * current leaseholder or 'null' (if the primary lease should be obtained
+   * by another client running in the foreground).
+   */
+  private tryBecomePrimary(): Promise<DbOwner | null> {
+    return this.simpleDb
+      .runTransaction('readwrite', ALL_STORES, txn => {
+        const metadataStore = clientMetadataStore(txn);
+        metadataStore.put(
+          new DbClientMetadata(this.clientKey, Date.now(), this.inForeground)
+        );
+        return this.tryElectPrimaryCandidate(txn).next(
+          electedPrimaryCandidate => {
+            if (this.isLocalClient(electedPrimaryCandidate)) {
+              return this.acquirePrimaryLease(txn);
+            }
+            return electedPrimaryCandidate;
+          }
+        );
+      })
+      .then(electedPrimaryCandidate => {
+        const isPrimary = this.isLocalClient(electedPrimaryCandidate);
+        if (this.isPrimary !== isPrimary) {
+          this.isPrimary = isPrimary;
+          this.primaryStateListener(this.isPrimary);
```

Comment:
I'm glad this is now happening on the AsyncQueue but I'm still concerned about component reentrancy. That is, SyncEngine could call into LocalStore which calls into IndexedDbPersistence.runTransaction() which hits this code which calls back into SyncEngine and now SyncEngine starts doing something "new" even though it was already in the middle of doing something else...  Generally we have not had components call back and forth into each other like that and allowing it will lead to complexity and bugs.

Perhaps we could solve it by just throwing a `queue.enqueue(() => ... )` around the notification (perhaps create a notifyPrimaryStateListener() helper method that does that).